src/Build/BackEnd/Node/OutOfProcServerNode.cs

```diff
@@ -0,0 +1,516 @@
+            {
+                if (disposing)
+                {
+                    Flush();
+                }
+
+                base.Dispose(disposing);
+            }
+
+            public override void Flush()
+            {
+                var sb = GetStringBuilder();
+                var captured = sb.ToString();
+                sb.Clear();
+                _writeCallback(captured);
+
+                base.Flush();
+            }
+
+            public override void Write(char value)
+            {
+                base.Write(value);
+                MaybeFlushCaptured(false);
+            }
+
+            public override void Write(char[] buffer, int index, int count)
+            {
+                base.Write(buffer, index, count);
+                MaybeFlushCaptured(buffer.SequenceEqual(CoreNewLine));
+            }
+
+            public override void Write(string? value)
+            {
+                if (value is null)
+                {
+                    return;
+                }
+
+                base.Write(value);
+                MaybeFlushCaptured(value.Contains(_newLineString));
+            }
+        }
+
+        /// <summary>
+        /// Handles the NodeBuildComplete packet.
+        /// </summary>
+        private void HandleNodeBuildComplete(NodeBuildComplete buildComplete)
+        {
+            _shutdownReason = buildComplete.PrepareForReuse ? NodeEngineShutdownReason.BuildCompleteReuse : NodeEngineShutdownReason.BuildComplete;
+            _shutdownEvent.Set();
+        }
+
+        internal sealed class ServerNamedMutex : IDisposable
+        {
+            public readonly Mutex _serverMutex;
+
+            public bool IsDisposed { get; private set; }
+
+            public bool IsLocked { get; private set; }
+
+            public ServerNamedMutex(string mutexName, out bool createdNew)
+            {
+                _serverMutex = new Mutex(
+                    initiallyOwned: true,
+                    name: mutexName,
+                    createdNew: out createdNew);
+
+                if (createdNew)
+                {
+                    IsLocked = true;
+                }
+            }
+
+            internal static ServerNamedMutex OpenOrCreateMutex(string name, out bool createdNew)
+            {
+                // TODO: verify it is not needed anymore
+                // if (PlatformInformation.IsRunningOnMono)
+                // {
+                //     return new ServerFileMutexPair(name, initiallyOwned: true, out createdNew);
+                // }
+                // else
+
+                return new ServerNamedMutex(name, out createdNew);
+            }
+
+            public static bool WasOpen(string mutexName)
+            {
+                try
+                {
+                    // we can't use TryOpenExisting as it is not supported in net3.5
+                    using var m = Mutex.OpenExisting(mutexName);
+                    return true;
+                }
+                catch
```

Comment:
nit: Should catch only the specific exception type(s) expected here (`WaitHandleCannotBeOpenedException` and `UnauthorizedAccessException`)?

Reply:
I moved this type from shared code and it will not be compiled while targeting net35, so I will use `TryOpenExisting`.